### PR TITLE
Fix #73 (`composer test` don't works for windows)

### LINKED ISSUE #73: `composer test` don't works for windows
Output:
```bash
>composer test
You are running composer with xdebug enabled. This has a major impact on runtime performance. See https://getcomposer.org/xdebug
> phpunit
Could not open input file: <my_project_folder>\vendor\bin\/../phpunit/phpunit/phpunit

```

### PR DESCRIPTION
**Pull request by Nayjest**
See [https://github.com/thephpleague/skeleton/issues/73](https://github.com/thephpleague/skeleton/issues/73)

This solution works for me, tested on Windows 7 and Debian.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -35,7 +35,7 @@
         }
     },
     "scripts": {
-        "test": "phpunit"
+        "test": "php vendor/phpunit/phpunit/phpunit"
```
Comment (shadowhand) [inline composer.json]: I think you want `php vendor/bin/phpunit`.

Comment (Nayjest) [inline composer.json]: No, vendor/bin/phpunit is not php file, it's bash script

Comment (Nayjest) [inline composer.json]: And I tried following way:
```javascript
"scripts": {
       "test": "vendor/bin/phpunit"
}
```
Doesn't works on windows too.

**Comment by shadowhand** [comment]
Honestly I'm not sure I agree with this change. It eschews best practices for the majority of users for a working configuration for a minority.

**Comment by Nayjest** [comment]
I researched more.
Issue reproduces when there is "!" character in project path.
Looks like this issue is more related to composer becouse it generates vendor/bin/phpunit.bat file that doesn't works correctly in some cases.

**Comment by Nayjest** [comment]
Upd: another reason to use this pull-request:

When script specified as "test": "phpunit" it uses default phpunit installed in system instead of phpunit from vendor folder. (I have path to phpunit in OS PATH environment variable)

@shadowhand what do you think about this?

**Comment by mAAdhaTTah** [comment]
> When script specified as "test": "phpunit" it uses default phpunit installed in system instead of phpunit from vendor folder. (I have path to phpunit in OS PATH environment variable)

I don't believe this is true. I believe `composer` sets the path to use the autoloaded bin files before the user-defined PATH.

> Looks like this issue is more related to composer becouse it generates vendor/bin/phpunit.bat file that doesn't works correctly in some cases.

I think PHPUnit has its own .bat files; I don't think composer generates them. This may be an upstream issue.

**Comment by shadowhand** [comment]
> I'm pretty sure composer sets the path to use the autoloaded bin files before the user-defined PATH.

It does https://getcomposer.org/doc/articles/scripts.md#writing-custom-commands:

> Note: Composer's bin-dir is pushed on top of the PATH so that binaries of dependencies are easily accessible as CLI commands when writing scripts.

**Comment by Nayjest** [comment]
> I don't believe this is true

I'm not deluding you. Maybe situation for unix is different, I don't know, but I double-checket it for Windows 7 right now.

#### 1. My composer.json:
(see "test": "phpunit")
```javascript
{
    "name": "nayjest/collection",
    "description": "Collections for PHP",
    "keywords": [
        "collection",
        "collections"
    ],
    "type": "library",
    "license": "MIT",
    "authors": [
        {
            "name": "Vitalii [Nayjest] Stepanenko",
            "email": "mail@vitaliy.in",
            "role": "Developer"
        }
    ],
    "require": {
        "php": ">=5.4.0",
        "nayjest/manipulator": "^2||^3",
        "evenement/evenement": "2.0.*"
    },
    "require-dev": {
        "phpunit/phpunit": "^4||^5",
        "danielstjules/stringy": "^1.9||^2"
    },
    "suggest": {
        "danielstjules/stringy": "Required to filter ObjectCollection by properties with getters"
    },
    "autoload": {
        "psr-4": {
            "Nayjest\\Collection\\": "src"
        }
    },
    "autoload-dev": {
        "psr-4": {
            "Nayjest\\Collection\\Test\\": "tests"
        }
    },
    "support": {
        "email": "mail@vitaliy.in"
    },
    "scripts": {
        "test": "phpunit"
    },
    "minimum-stability": "stable"
}
```
#### 2. `composer test ` output

```
C:\Users\Vitaliy Stepanenko\Dropbox\!CODE\lib\collection>composer test
You are running composer with xdebug enabled. This has a major impact on runtime
 performance. See https://getcomposer.org/xdebug
> phpunit

C:\Users\Vitaliy Stepanenko\Dropbox\!CODE\lib\collection>php D:\bin\phpunit.phar


Warning: Missing argument 3 for PHPUnit_TextUI_TestRunner::doRun(), called in ph
ar://D:/bin/phpunit.phar/phpunit/TextUI/Command.php on line 152 and defined in C
:\Users\Vitaliy Stepanenko\Dropbox\!CODE\lib\collection\vendor\phpunit\phpunit\s
rc\TextUI\TestRunner.php on line 148

Call Stack:
    0.0150     570584   1. {main}() D:\bin\phpunit.phar:0
    0.0160     800488   2. PHPUnit_TextUI_Command::main() D:\bin\phpunit.phar:72
2
    0.0160     801128   3. PHPUnit_TextUI_Command->run() phar://D:/bin/phpunit.p
har/phpunit/TextUI/Command.php:104
    0.0860    3745512   4. PHPUnit_TextUI_TestRunner->doRun() phar://D:/bin/phpu
nit.phar/phpunit/TextUI/Command.php:152

PHPUnit 5.2.9 by Sebastian Bergmann and contributors.

........................................                         40 / 40 (100%)

Time: 1.6 seconds, Memory: 12.50Mb

OK (40 tests, 85 assertions)

Generating code coverage report in Clover XML format ... done

Generating code coverage report in HTML format ... done

Notice: Undefined variable: exit in C:\Users\Vitaliy Stepanenko\Dropbox\!CODE\li
b\collection\vendor\phpunit\phpunit\src\TextUI\TestRunner.php on line 562

Call Stack:
    0.0150     570584   1. {main}() D:\bin\phpunit.phar:0
    0.0160     800488   2. PHPUnit_TextUI_Command::main() D:\bin\phpunit.phar:72
2
    0.0160     801128   3. PHPUnit_TextUI_Command->run() phar://D:/bin/phpunit.p
har/phpunit/TextUI/Command.php:104
    0.0860    3745512   4. PHPUnit_TextUI_TestRunner->doRun() phar://D:/bin/phpu
nit.phar/phpunit/TextUI/Command.php:152

Для продолжения нажмите любую клавишу . . .

C:\Users\Vitaliy Stepanenko\Dropbox\!CODE\lib\collection>
```

**Now please pay attention to following line:**
```
Warning: Missing argument 3 for PHPUnit_TextUI_TestRunner::doRun(), called in ph
ar://D:/bin/phpunit.phar/phpunit/TextUI/Command.php on line 152 and defined in C
:\Users\Vitaliy Stepanenko\Dropbox\!CODE\lib\collection\vendor\phpunit\phpunit\s
rc\TextUI\TestRunner.php on line 148
```

`D:/bin/phpunit.phar ` located inside `D:/bin` which is placed to OS PATH environmant variable on my system.

**Comment by Nayjest** [comment]
Upd: I updated composer before this test.

**Comment by Nayjest** [comment]
> It does https://getcomposer.org/doc/articles/scripts.md#writing-custom-commands:

Obviously composer does not works as described.

When I trying to run tests using composer downloaded as .phar right now and running it via `phph composer.phar test`, I get:

```php
php composer.phar test
> phpunit
Could not open input file: C:\Users\Vitaliy Stepanenko\Dropbox\CODE\lib\collection\vendor\bin\/../phpunit/phpunit/phpunit
```

Problem is that it removed "!" from "!CODE" folder in path.

And when I trying to use composer installed to my system, it uses phpunit installed to my system instead of vendor/bin/phpunit.

Proof that composer installed to my system isn't outdated:

```
C:\Users\Vitaliy Stepanenko\Dropbox\!CODE>composer -V
Composer version 1.0-dev (ef6b7cdf12dc0331ed1aa9fde8be04e18c8e07e2) 2016-03-01 1
4:37:55
```

I will submit this to composer issue tracker.

Regarding "thephpleague/skeleton", here is two options:

1. Use command from this pull-request, that works everywhere (I hope so, tested on my windows & debian, suggest, you need to test it too)
2. Leave as is and wait when composer issue will be fixed

**Comment by mAAdhaTTah** [comment]
Ok, I believe you, but if it's an upstream issue, it should be fixed by composer and/or PHPUnit (if there's a problem with its .bat file), not here. I would recommend opening an issue with them with the same details you described above. They should resolve this problem for all Windows users.

**Comment by Nayjest** [comment]
Created issue in composer issue tracker:
https://github.com/composer/composer/issues/4990

File: composer.json (hunk 1 above)
Comment (sanmai) [inline composer.json]: `vendor/bin/phpunit` is a PHP file but without `.php` extension

     #!/usr/bin/env php
      <?php

As [in the docs](http://php.net/manual/en/features.commandline.usage.php): "The special shebang first line for Unix does no harm on Windows"

**Comment by Nayjest** [comment]
Issue fixed in composer

**Pull request closed without merging**